File: java/src/stressor/pool/Pool.java
package stressor.pool;

import java.util.ArrayDeque;
import java.util.Deque;

import stressor.random.Random;
import stressor.tracker.Tracker;
import stressor.utils.ArgsParser.Policy;

public class Pool {
    public static class Block {
        public byte[] data;
        public long ttl;
        public long ttl_org;

        Block(int size) {
            this(size, -1);
        }

        Block(int size, long ttl) {
            this.data = new byte[size];
            this.ttl = ttl;
            this.ttl_org = ttl;
            Tracker.get().addAlloc(size);
        }

        public void tick() {
            if (this.ttl > 0) {
                this.ttl--;
            }
        }

        public boolean alive() {
            return (this.ttl < 0) || (this.ttl > 0);
        }
    }

    public final Deque<Pool.Block> blocks;
    public final int capacity;

    public Pool(int capacity) {
        this.capacity = capacity;
        this.blocks = new ArrayDeque<>(capacity);
    }

    public void deinit() {
        while (!blocks.isEmpty()) {
            var b = blocks.removeLast();
            Tracker.get().removeAlloc(b.data.length);
        }
    }

    public int count() {
        return this.blocks.size();
    }

    public void addBlock(int size) {
        blocks.addLast(new Block(size));
    }

    public void addBlock(int size, long ttl) {
        blocks.addLast(new Block(size, ttl));
    }

    public void delBlock(Policy policy) {
        if (blocks.size() == 0) {
            return;
        }
        
        switch (policy) {
            case LIFO: {
                var b = blocks.removeLast();
                Tracker.get().removeAlloc(b.data.length);
            }
                return;
            case FIFO: {
                var b = blocks.removeFirst();
                Tracker.get().removeAlloc(b.data.length);
            }
                return;
            case RANDOM: {
                int idx = (int) Random.get().uniform(0, blocks.size());
                var it = blocks.iterator();
                int i = 0;
                while (it.hasNext()) {
                    var b = it.next();
                    if (i == idx) {
                        Tracker.get().removeAlloc(b.data.length);
                        it.remove();
                        break;
                    }
                    i++;
                }
                return;
            }
            case BIG_FIRST: {
                var el = blocks.stream().max((lhs, rhs) -> lhs.data.length - rhs.data.length);
                if (el.isPresent()) {
                    Tracker.get().removeAlloc(el.get().data.length);
                    blocks.remove(el.get());
                }
            }
                return;
            case SMALL_FIRST:
                var el = blocks.stream().min((lhs, rhs) -> lhs.data.length - rhs.data.length);
                if (el.isPresent()) {
                    Tracker.get().removeAlloc(el.get().data.length);
                    blocks.remove(el.get());
                }
                return;
            case NEVER:
                return;
        }

        throw new RuntimeException("Unreachable(del_block)");
    }

    public void updateAndPrune() {
        var it = this.blocks.iterator();

        while (it.hasNext()) {
            var b = it.next();
            b.tick();

            if (!b.alive()) {
                Tracker.get().removeAlloc(b.data.length);
                it.remove();
            }
        }
    }
}

File: rust/src/pool.rs
use std::collections::VecDeque;

use crate::random::Random;
use crate::tracker::GLOBAL_TRACKER;
use crate::utils::args_parser::Policy;

pub struct Block {
    data: Vec<u8>,
    ttl: isize,
    ttl_org: isize,
}

impl Drop for Block {
    fn drop(&mut self) {
        unsafe {
            GLOBAL_TRACKER.remove_alloc(self.data.capacity());
        }
    }
}

impl Block {
    pub fn new(size: usize) -> Self {
        Self::with_ttl(size, -1)
    }

    pub fn with_ttl(size: usize, ttl: isize) -> Self {
        unsafe {
            GLOBAL_TRACKER.add_alloc(size);
        }
        Self {
            data: vec![0; size],
            ttl,
            ttl_org: ttl,
        }
    }

    #[inline(always)]
    pub fn tick(&mut self) {
        if self.ttl > 0 {
            self.ttl -= 1;
        }
    }

    #[inline(always)]
    pub const fn alive(&self) -> bool {
        self.ttl < 0 || self.ttl > 0
    }

    #[inline(always)]
    pub const fn ttl_org(&self) -> isize {
        self.ttl_org
    }
}

pub struct Pool {
    blocks: VecDeque<Block>,
    pub capacity: usize,
}

impl Pool {
    pub fn new(capacity: usize) -> Self {
        Self {
            blocks: VecDeque::with_capacity(capacity),
            capacity: capacity,
        }
    }

    pub fn add_block(&mut self, size: usize) {
        if self.blocks.len() >= self.capacity {
            return;
        }
        self.blocks.push_back(Block::new(size));
    }

    pub fn add_block_with_ttl(&mut self, size: usize, ttl: isize) {
        if self.blocks.len() >= self.capacity {
            return;
        }
        self.blocks.push_back(Block::with_ttl(size, ttl));
    }

    pub fn del_block(&mut self, policy: Policy, rng: &mut Random) {
        match policy {
            Policy::Lifo => self.blocks.pop_back(),
            Policy::Fifo => self.blocks.pop_front(),
            Policy::Random => self.blocks.remove(rng.uniform(0, self.blocks.len())),
            Policy::BigFirst => self.blocks.remove(
                self.blocks
                    .iter()
                    .enumerate()
                    .max_by_key(|(_, b)| b.data.capacity())
                    .map(|(i, _)| i)
                    .unwrap_or(0),
            ),
            Policy::SmallFirst => self.blocks.remove(
                self.blocks
                    .iter()
                    .enumerate()
                    .min_by_key(|(_, b)| b.data.capacity())
                    .map(|(i, _)| i)
                    .unwrap_or(0),
            ),
            Policy::Never => None,
        };
    }

    pub fn update_and_prune(&mut self) {
        let mut i = 0usize;
        while i < self.blocks.len() {
            self.blocks[i].tick();

            if !self.blocks[i].alive() {
                self.blocks.remove(i);
            } else {
                i += 1;
            }
        }
    }

    pub fn count(&self) -> usize {
        self.blocks.len()
    }
}
